File: platforms/scada/packages/renderer/src/types/project.ts
/**
 * 工程类型定义
 * 定义 Scada 工程管理功能的所有数据结构
 */

/**
 * 工程分类枚举（内置分类 + 自定义分类）
 */
export enum ProjectCategory {
  CATEGORY_1 = '分类1',
  CATEGORY_2 = '分类2',
  CUSTOM = '新建分类...'
}

/**
 * 运行平台枚举（设备型号）
 * 注意：平台列表从后端API获取，此枚举仅用于TypeScript类型检查
 */
export enum HardwarePlatform {
  BOX1 = 'BOX1',     // 标准BOX型号
  HMI01 = 'HMI01',   // HMI触摸屏
  TBOX1 = 'TBOX1',   // 网关盒子
  // 未来可能添加的平台
  // BOX2 = 'BOX2',
  // HMI02 = 'HMI02',
}

/**
 * 硬件平台接口（从后端获取）
 */
export interface HardwarePlatformConfig {
  id: string              // 平台唯一标识
  name: string            // 平台显示名称
  type: PlatformType      // 平台类型
  resolution?: string     // 屏幕分辨率
  enabled: boolean        // 是否启用
}

/**
 * 平台类型枚举
 */
export enum PlatformType {
  BOX = 'box',
  HMI = 'hmi',
  GATEWAY = 'gateway'
}

/**
 * 工程元数据
 */
export interface ProjectMetadata {
  /** 工程名称 */
  name: string
  /** 工程作者 */
  author?: string
  /** 工程描述 */
  description?: string
  /** 工程分类 */
  category: string
  /** 运行平台（设备型号） */
  platform: string
  /** 创建时间 (ISO 8601) */
  createdAt: string
  /** 修改时间 (ISO 8601) */
  updatedAt: string
}

/**
 * 工程安全配置
 */
export interface ProjectSecurity {
  /** 是否加密 */
  encrypted: boolean
  /** 密码哈希 (bcrypt, 仅加密工程有) */
  passwordHash?: string
  /** 设备指纹 (可选) */
  deviceBinding?: string
  /** 文件签名 (HMAC-SHA256) */
  fileSignature: string
  /** KEK版本 (双重加密机制) */
  kekVersion?: string
  /** 数据加密密钥 (双重加密机制) */
  dek?: {
    /** 用户密码加密的DEK */
    userEncrypted: string
    /** 官方KEK加密的DEK备份 */
    officialEncrypted: string
  }
}

/**
 * 画布配置
 */
export interface CanvasConfig {
  /** 画布宽度 */
  width: number
  /** 画布高度 */
  height: number
  /** 背景颜色 */
  backgroundColor?: string
}

/**
 * 组件定义
 */
export interface Component {
  /** 组件ID */
  id: string
  /** 组件类型 */
  type: string
  /** X坐标 */
  x: number
  /** Y坐标 */
  y: number
  /** 宽度 */
  width: number
  /** 高度 */
  height: number
  /** 组件属性 */
  properties: Record<string, any>
  /** 数据绑定 */
  dataBindings?: DataBinding[]
}

/**
 * 数据绑定
 */
export interface DataBinding {
  /** 组件ID */
  componentId: string
  /** 属性名 */
  property: string
  /** 数据源 */
  source: string
}

/**
 * 工程主数据结构
 */
export interface Project {
  /** 文件格式版本 */
  version: string
  /** 工程唯一标识 (UUID v4) */
  projectId: string
  /** 工程元数据 */
  metadata: ProjectMetadata
  /** 安全配置 */
  security: ProjectSecurity
  /** 画布配置 */
  canvas: CanvasConfig
  /** 组件列表 */
  components: Component[]
  /** 加密内容 (仅加密工程有) */
  encryptedContent?: string
}

/**
 * 最近工程数据
 */
export interface RecentProject {
  /** 工程唯一标识 */
  projectId: string
  /** 工程名称 */
  name: string
  /** 工程分类 */
  category?: string
  /** 工程文件路径 */
  filePath: string
  /** 最后打开时间 (相对时间字符串, 如 "2小时前") */
  lastOpened: string
  /** 最后打开时间 (Date对象) */
  lastOpenedDate: Date
  /** 是否加密 */
  isEncrypted: boolean
  /** 创建时间 */
  createdAt: string
}

/**
 * 新建工程表单数据
 */
export interface NewProjectFormData {
  /** 工程名称 */
  name: string
  /** 工程作者 */
  author?: string
  /** 工程描述 */
  description?: string
  /** 工程分类 */
  category: string
  /** 运行平台 */
  platform: string
  /** 工程加密 */
  encrypted: boolean
  /** 密码 (加密工程需要) */
  password?: string
  /** 确认密码 */
  confirmPassword?: string
  /** 保存位置 */
  savePath: string
}

/**
 * 打开工程请求数据
 */
export interface OpenProjectRequest {
  /** 工程文件路径 */
  filePath: string
  /** 密码 (加密工程需要) */
  password?: string
}

/**
 * API响应基础结构
 */
export interface ApiResponse<T = any> {
  /** 是否成功 */
  success: boolean
  /** 数据 */
  data?: T
  /** 错误码 */
  error?: string
  /** 错误信息 */
  message?: string
}

/**
 * 创建工程响应
 */
export interface CreateProjectResponse {
  /** 工程ID */
  projectId: string
  /** 工程文件路径 */
  filePath: string
}

/**
 * 打开工程响应
 */
export interface OpenProjectResponse {
  /** 工程数据 */
  project: Project
}

File: platforms/scada/backend/internal/model/project.go
/**
 * Scada 工程数据模型
 * 定义工程管理功能的所有数据结构
 */

package model

import (
	"errors"
	"time"
)

// 错误定义
var (
	ErrInvalidPassword  = errors.New("invalid password")
	ErrInvalidSignature = errors.New("invalid signature")
)

/**
 * 工程分类
 */
type ProjectCategory string

const (
	Category1 ProjectCategory = "分类1"
	Category2 ProjectCategory = "分类2"
	Custom    ProjectCategory = "自定义分类"
)

/**
 * 硬件平台
 */
type HardwarePlatform string

const (
	HMIModel1    HardwarePlatform = "HMI型号1"
	HMIModel2    HardwarePlatform = "HMI型号2"
	GatewayModel1 HardwarePlatform = "网关型号1"
)

/**
 * 工程元数据
 */
type ProjectMetadata struct {
	Name        string          `json:"name"`
	Author      string          `json:"author,omitempty"`
	Description string          `json:"description,omitempty"`
	Category    string          `json:"category"`
	Platform    HardwarePlatform `json:"platform"`
	CreatedAt   string          `json:"createdAt"`
	UpdatedAt   string          `json:"updatedAt"`
}

/**
 * 工程安全配置
 */
type ProjectSecurity struct {
	Encrypted     bool   `json:"encrypted"`
	Password      string `json:"password,omitempty"`      // 临时字段，用于接收用户密码（创建/打开时），不序列化到文件
	PasswordHash  string `json:"passwordHash,omitempty"`
	DeviceBinding string `json:"deviceBinding,omitempty"`
	FileSignature string `json:"fileSignature"`
	KEKVersion    string `json:"kekVersion,omitempty"`
	// 双重加密机制 - DEK
	UserEncryptedDEK       string `json:"userEncrypted,omitempty"`
	OfficialEncryptedDEK   string `json:"officialEncrypted,omitempty"`
}

/**
 * 画布配置
 */
type CanvasConfig struct {
	Width           int     `json:"width"`
	Height          int     `json:"height"`
	BackgroundColor string  `json:"backgroundColor,omitempty"`
}

/**
 * 数据绑定
 */
type DataBinding struct {
	ComponentID string `json:"componentId"`
	Property    string `json:"property"`
	Source      string `json:"source"`
}

/**
 * 组件定义
 */
type Component struct {
	ID            string         `json:"id"`
	Type          string         `json:"type"`
	X             int            `json:"x"`
	Y             int            `json:"y"`
	Width         int            `json:"width"`
	Height        int            `json:"height"`
	Properties    map[string]any `json:"properties"`
	DataBindings  []DataBinding  `json:"dataBindings,omitempty"`
}

/**
 * 工程主数据结构
 */
type Project struct {
	Version          string             `json:"version"`
	ProjectID        string             `json:"projectId"`
	Metadata         ProjectMetadata    `json:"metadata"`
	Security         ProjectSecurity    `json:"security"`
	Canvas           CanvasConfig       `json:"canvas"`
	Components       []Component        `json:"components"`
	EncryptedContent string             `json:"encryptedContent,omitempty"`
}

/**
 * 最近工程（数据库模型）
 */
type RecentProject struct {
	ID          uint      `gorm:"primaryKey;autoIncrement" json:"id"`
	ProjectID   string    `gorm:"uniqueIndex;not null" json:"projectId"`
	Name        string    `gorm:"not null" json:"name"`
	Category    string    `json:"category,omitempty"`
	FilePath    string    `gorm:"not null" json:"filePath"`
	LastOpened  time.Time `gorm:"not null;index" json:"lastOpened"`
	IsEncrypted bool      `gorm:"default:false" json:"isEncrypted"`
	CreatedAt   time.Time `gorm:"autoCreateTime" json:"createdAt"`
}

/**
 * 自定义分类（数据库模型）
 */
type CustomCategory struct {
	CategoryID string    `gorm:"primaryKey" json:"categoryId"`
	Name       string    `gorm:"not null" json:"name"`
	Color      string    `gorm:"default:'#2196F3'" json:"color"`
	CreatedAt  time.Time `gorm:"autoCreateTime" json:"createdAt"`
}

/**
 * 应用配置（数据库模型）
 */
type AppConfig struct {
	Key       string    `gorm:"primaryKey" json:"key"`
	Value     string    `gorm:"not null" json:"value"`
	UpdatedAt time.Time `gorm:"autoUpdateTime" json:"updatedAt"`
}

/**
 * 用户偏好（数据库模型）
 */
type UserPreference struct {
	Key       string    `gorm:"primaryKey" json:"key"`
	Value     string    `gorm:"not null" json:"value"`
	UpdatedAt time.Time `gorm:"autoUpdateTime" json:"updatedAt"`
}

/**
 * 审计日志（数据库模型）
 */
type AuditLog struct {
	ID        uint      `gorm:"primaryKey;autoIncrement" json:"id"`
	Timestamp time.Time `gorm:"not null;index" json:"timestamp"`
	Operator  string    `gorm:"not null;index" json:"operator"`
	Operation string    `gorm:"not null" json:"operation"`
	TargetID  string    `json:"targetId,omitempty"`
	Details   string    `json:"details,omitempty"`
	Result    string    `json:"result,omitempty"`
	CreatedAt time.Time `gorm:"autoCreateTime" json:"createdAt"`
}

/**
 * KEK版本管理（数据库模型）
 */
type KEKVersion struct {
	VersionID   string    `gorm:"primaryKey" json:"versionId"`
	Version     string    `gorm:"not null" json:"version"`
	CreatedAt   time.Time `gorm:"not null" json:"createdAt"`
	RotatedAt   *time.Time `json:"rotatedAt,omitempty"`
	Status      string    `gorm:"default:'active'" json:"status"` // active, rotated, expired
}

/**
 * 创建工程请求
 */
type CreateProjectRequest struct {
	Metadata  ProjectMetadata `json:"metadata"`
	Security  ProjectSecurity `json:"security"`
	SavePath  string          `json:"savePath"`
}

/**
 * 打开工程请求
 */
type OpenProjectRequest struct {
	FilePath string `json:"filePath"`
	Password string `json:"password,omitempty"`
}

/**
 * 验证密码请求
 */
type ValidatePasswordRequest struct {
	FilePath string `json:"filePath"`
	Password string `json:"password"`
}

/**
 * 保存工程请求
 */
type SaveProjectRequest struct {
	Project Project `json:"project"`
}
